hipcheck/src/policy_exprs/env.rs

```diff
@@ -23,14 +32,274 @@ pub struct Env<'parent> {
 #[derive(Clone)]
 pub enum Binding {
 	/// A function.
-	Fn(Op),
+	Fn(FunctionDef),
 
 	/// A primitive value.
 	Var(Primitive),
 }
 
-/// Helper type for operation function pointer.
-type Op = fn(&Env, &[Expr]) -> Result<Expr>;
+// Ensure that type of array elements is valid with a lambda
+fn ty_check_higher_order_lambda(
+	l_ty: &FunctionType,
+	arr_ty: &ExprArrayType,
+) -> Result<ReturnableType> {
+	if let Some(arr_elt_ty) = arr_ty {
+		// Copy the lambda function type, replace ident with arr_elt_ty
+		let mut try_l_ty = l_ty.clone();
+		let first_arg = try_l_ty.arg_tys.get_mut(0).ok_or(Error::NotEnoughArgs {
+			name: "".to_owned(),
+			expected: 1,
+			given: 0,
+		})?;
+		*first_arg = Type::Primitive(*arr_elt_ty);
+		// If this returns error, means array type was incorrect for lambda
+		try_l_ty.get_return_type()
+	} else {
+		Ok(ReturnableType::Unknown)
+	}
+}
+
+fn ty_filter(args: &[Type]) -> Result<ReturnableType> {
+	let arr_ty = expect_array_at(args, 1)?;
+	let wrapped_l_ty = args.first().unwrap();
+	let Type::Lambda(l_ty) = wrapped_l_ty else {
+		return Err(Error::BadFuncArgType {
+			name: "".to_owned(),
+			idx: 0,
+			expected: "a lambda".to_owned(),
+			got: wrapped_l_ty.clone(),
+		});
+	};
+	let res_ty = ty_check_higher_order_lambda(l_ty, &arr_ty)?;
+	match res_ty {
+		ReturnableType::Primitive(PrimitiveType::Bool) | ReturnableType::Unknown => {
+			Ok(ReturnableType::Array(arr_ty))
+		}
+		a => Err(Error::BadFuncArgType {
+			name: "".to_owned(),
+			idx: 0,
+			expected: "a bool-returning lambda".to_owned(),
+			got: Type::Lambda(l_ty.clone()),
+		}),
+	}
+}
+
+fn ty_higher_order_bool_fn(args: &[Type]) -> Result<ReturnableType> {
+	let arr_ty = expect_array_at(args, 1)?;
+	let wrapped_l_ty = args.first().unwrap();
+	let Type::Lambda(l_ty) = wrapped_l_ty else {
+		return Err(Error::BadFuncArgType {
+			name: "".to_owned(),
+			idx: 0,
+			expected: "a lambda".to_owned(),
+			got: wrapped_l_ty.clone(),
+		});
+	};
+	let res_ty = ty_check_higher_order_lambda(l_ty, &arr_ty)?;
+	match res_ty {
+		ReturnableType::Primitive(PrimitiveType::Bool) | ReturnableType::Unknown => {
+			Ok(ReturnableType::Primitive(PrimitiveType::Bool))
+		}
+		a => Err(Error::BadFuncArgType {
+			name: "".to_owned(),
+			idx: 0,
+			expected: "a bool-returning lambda".to_owned(),
+			got: Type::Lambda(l_ty.clone()),
+		}),
+	}
+}
+
+// Type of dynamic function is dependent on first arg
+fn ty_inherit_first(args: &[Type]) -> Result<ReturnableType> {
+	args.first().unwrap().try_into()
+}
+
+fn ty_from_first_arr(args: &[Type]) -> Result<ReturnableType> {
+	let arr_ty = expect_array_at(args, 0)?;
```

Comment:
Why do `ty_filter`, `ty_higher_order_bool_fn`, and `ty_foreach` use `expect_array_at(args, 1)`, but `ty_from_first_arr` uses an index of `0`?

After thinking more: It seems to be because those functions expect a function type to be the first argument (at index `0`).

These feel like magic numbers. Maybe it would be enough to comment these functions to explain what is expected to be at these indices.

It might also help to reorder the code lines in those other functions, to make it clearer that the lambda comes first, before the array starting at index `1`.

Reply:
Added comments to explain positioning expectations.